File: src/Benchmarks/RosettaCode/RosettaCode/RosettaCode.py
## setup
import time
mintrials = 10

def timeit(name):
    tmin = float('inf')
    for i in range(mintrials):
        t = time.time()
        f = main()
        t = time.time()-t
        if t < tmin: tmin = t
    print(name + "\t" + str(tmin*1000))

## benchmarks here

def RK4(f):
    return lambda t, y, dt: (
            lambda dy1: (
            lambda dy2: (
            lambda dy3: (
            lambda dy4: (dy1 + 2*dy2 + 2*dy3 + dy4)/6
            )( dt * f( t + dt  , y + dy3   ) )
	    )( dt * f( t + dt/2, y + dy2/2 ) )
	    )( dt * f( t + dt/2, y + dy1/2 ) )
	    )( dt * f( t       , y         ) )


from math import sqrt

def main():
    dy = RK4(lambda t, y: t*sqrt(y))
    t, y, dt = 0., 1., .1
    for i in range(0,1001):
        t, y = t + dt, y + dy( t, y, dt )
    return y

assert (main() - 6.280048524572245e6) < 1e-12
timeit("Runge-Kutta method")

File: src/Benchmarks/RosettaCode/RosettaCode/RosettaCode.jl
include(joinpath(@__DIR__,"../../../timeit.jl"))

function main()
    f(x, y) = x * sqrt(y)
    theoric(t) = (t ^ 2 + 4.0) ^ 2 / 16.0

    rk4(f) = (t, y, δt) ->  # 1st (result) lambda
             ((δy1) ->      # 2nd lambda
             ((δy2) ->      # 3rd lambda
             ((δy3) ->      # 4th lambda
             ((δy4) -> ( δy1 + 2δy2 + 2δy3 + δy4 ) / 6 # 5th and deepest lambda: calc y_{n+1}
             )(δt * f(t + δt, y + δy3))         # calc δy₄
             )(δt * f(t + δt / 2, y + δy2 / 2)) # calc δy₃
             )(δt * f(t + δt / 2, y + δy1 / 2)) # calc δy₂
             )(δt * f(t, y))                    # calc δy₁

    δy = rk4(f)
    t₀, δt = 0.0, 0.1
    y₀ = 1.0

    t, y = t₀, y₀
    for i=1:1001
        y += δy(t, y, δt)
        t += δt
    end
    y
end

@assert (main() - 6.280048524572245e6) < 1e-12
timeit("Runge-Kutta method",main)
